src/details/ArborX_DetailsCartesianGrid.hpp

```diff
@@ -0,0 +1,109 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILSCARTESIANGRID_HPP
+#define ARBORX_DETAILSCARTESIANGRID_HPP
+
+#include <ArborX_Exception.hpp>
+#include <ArborX_GeometryTraits.hpp>
+
+#include <Kokkos_Macros.hpp>
+
+#include <cassert>
+
+namespace ArborX
+{
+namespace Details
+{
+
+template <class Box>
```

Comment:
Why templating on `Box` rather than the number of dimensions?

Reply:
We want to be able to return a box corresponding to a cell index. If we just provide dimension, it is unclear what box type is and it won't work with both old and new geometries.

Reply:
I still think we should template the grid over the dimension, not the box type

Reply:
Well, this PR would then have to wait until we merge all others, because `cellBox` will have to return the experimental box, which right now cannot be used in DBSCAN.